File: db/decision.go
// Package db 決策引擎（腦）：Decide(state, context) -> Intent。
// 對齊 docs/reference/奇點決策引擎架構.md、奇點馬斯洛需求系統.md、討論 002。
// 引擎只產生意圖，不寫 DB、不發送；由主迴圈或 TravelerManager 依 Intent 驅動移動或插座。
package db

import (
	"database/sql"
	"math/rand"
	"time"
)

// 生存層觸發的鎂閾值；低於此值視為生存未滿足，優先求職／乞討／採集。
const SurvivalLineMg = 50

// socialDecaySec 社交需求衰減週期（秒）：超過此時間未交談，社交緊迫度漸升至 1.0。
const socialDecaySec = 7200 // 約兩小時真實時間

// IntentType 意圖類型：對應插座語義（決策 002）與討論 002 行為。
type IntentType string

const (
	IntentSeekJob   IntentType = "seek_job"   // 求職：前往有場所的房間，路過或撮合寫 assignment
	IntentBeg       IntentType = "beg"        // 乞討：gate/social 執行 Beg
	IntentTrade     IntentType = "trade"      // 賣貨：行腳商／帶貨兜售
	IntentGather    IntentType = "gather"     // 採集：wilderness/outdoor 執行 Gather
	IntentWander    IntentType = "wander"     // 閒逛：無明確目標，隨機移動或停留
	IntentWork      IntentType = "work"       // 上班：有排班時由 TravelerManager 處理，腦不產出
	IntentIdle      IntentType = "idle"       // 閒置：在 social 等釋放 Idle 敘事
	IntentSocialize IntentType = "socialize"  // 社交：前往 tavern/inn/social，釋放 idle 對話、調整心境
)

// Intent 決策輸出：單一意圖與可選目標房間（由主迴圈／TravelerManager 執行移動或插座）。
type Intent struct {
	Type         IntentType
	TargetRoomID string // 可為空：表示在當前房間執行插座，或由上層自行選目標
}

// DecisionState 決策輸入：NPC 自身狀態（馬斯洛掃描用）。
type DecisionState struct {
	EntityID       string
	Magnesium      int
	HasAssignment  bool
	Personality    Personality
	HasPersonality bool
	CurrentRoomID  string
	LastTalkAt     int64 // Unix 時間戳；0 = 從未交談；社交需求層用
	Disposition    int   // 心境值 [-100, +100]；影響社交與乞討意圖強度
}

// DecisionContext 決策輸入：世界情境（場所、房間標籤、鄰近可達房間）。
type DecisionContext struct {
	RoomTags                 []string // 當前房間標籤
	VenueRoomIDsInRange      []string // 距離內屬於某場所的房間（求職目標）
	SocialOrGateInRange      []string // 距離內帶 social / gate 的房間（乞討）
	WildernessOutdoorInRange []string // 距離內帶 wilderness / outdoor 的房間（採集）
	TavernInRange            []string // 距離內帶 tavern / inn 的房間（社交意圖目標）
}

// BuildDecisionState 從 DB 組裝 NPC 的決策狀態。
func BuildDecisionState(db *sql.DB, entityID string) (DecisionState, error) {
	var s DecisionState
	s.EntityID = entityID
	c, err := GetEntity(db, entityID)
	if err != nil || c == nil {
		return s, err
	}
	s.Magnesium = c.Magnesium
	s.Disposition = c.Disposition
	assignments, _ := GetAssignmentsForEntity(db, entityID)
	s.HasAssignment = len(assignments) > 0
	if c.SoulSeed != nil {
		s.Personality = ExpandSoulSeedToPersonality(*c.SoulSeed)
		s.HasPersonality = true
	}
	s.CurrentRoomID, _ = GetEntityRoom(db, entityID)
	// 最近一筆 EvtTalk 事件時間戳作為社交衰減起點
	for _, e := range GetRecentEvents(entityID, 10) {
		if e.EventType == EvtTalk {
			s.LastTalkAt = e.At
			break
		}
	}
	return s, nil
}

// BuildDecisionContext 從 RoomGraph 與 DB 組裝當前房間周邊情境（maxDist 步內）。
func BuildDecisionContext(db *sql.DB, g *RoomGraph, currentRoomID string, maxDist int) DecisionContext {
	if maxDist <= 0 {
		maxDist = 20
	}
	var ctx DecisionContext
	ctx.RoomTags = g.RoomTags(currentRoomID)
	venueRooms, _ := GetAllVenueRoomIDs(db)
	ctx.VenueRoomIDsInRange = filterRoomsWithinDist(g, currentRoomID, venueRooms, maxDist)
	ctx.SocialOrGateInRange = g.FindRoomsWithinDist(currentRoomID, []string{"social", "gate"}, maxDist)
	ctx.WildernessOutdoorInRange = g.FindRoomsWithinDist(currentRoomID, []string{"wilderness", "outdoor"}, maxDist)
	ctx.TavernInRange = g.FindRoomsWithinDist(currentRoomID, []string{"tavern", "inn"}, maxDist)
	return ctx
}

// filterRoomsWithinDist 從候選房間中篩出在 origin 的 maxDist 步內者。
func filterRoomsWithinDist(g *RoomGraph, origin string, candidates []string, maxDist int) []string {
	if len(candidates) == 0 || maxDist <= 0 {
		return nil
	}
	var out []string
	for _, c := range candidates {
		if c == origin {
			out = append(out, c)
			continue
		}
		path := g.FindPath(origin, c)
		if len(path) > 0 && len(path) <= maxDist {
			out = append(out, c)
		}
	}
	return out
}

// hasTag 判斷房間標籤是否包含任一給定 tag。
func hasTag(tags []string, want ...string) bool {
	for _, t := range tags {
		for _, w := range want {
			if t == w {
				return true
			}
		}
	}
	return false
}

// candidate 候選行為與權重（供性格加權選擇）。
type candidate struct {
	intent Intent
	weight float64
}

// personalityWeightedSelect 依性格加權隨機選一意圖。Boldness 高→Gather/Trade↑ Beg↓；Orderliness 高→SeekJob↑；Sensitivity 高→Beg↑。
func personalityWeightedSelect(candidates []candidate, p Personality, hasPers bool) Intent {
	if len(candidates) == 0 {
		return Intent{Type: IntentWander}
	}
	if len(candidates) == 1 || !hasPers {
		return candidates[0].intent
	}
	for i := range candidates {
		switch candidates[i].intent.Type {
		case IntentSeekJob:
			candidates[i].weight *= 0.5 + p.Orderliness
		case IntentBeg:
			candidates[i].weight *= 1.5 - p.Boldness
			candidates[i].weight *= 0.5 + p.Sensitivity
		case IntentGather:
			candidates[i].weight *= 0.5 + p.Boldness
			candidates[i].weight *= 1.5 - p.Orderliness
		case IntentTrade:
			candidates[i].weight *= 0.5 + p.Boldness
		case IntentWander:
			candidates[i].weight *= 1.5 - p.Orderliness
		}
	}
	total := 0.0
	for _, c := range candidates {
		total += c.weight
	}
	if total <= 0 {
		return candidates[0].intent
	}
	r := rand.Float64() * total
	for _, c := range candidates {
		r -= c.weight
		if r <= 0 {
			return c.intent
		}
	}
	return candidates[len(candidates)-1].intent
}

// Decide V2：評分引擎。各需求層同時計算緊迫度分數，性格三軸直接影響各意圖權重，
// 加權隨機選出最終意圖；取代 V1 固定 if-else 優先序。
func Decide(state DecisionState, ctx DecisionContext) Intent {
	survival := urgencySurvival(state.Magnesium)
	stability := urgencyStability(state.HasAssignment)
	social := urgencySocial(state.LastTalkAt, state.Disposition)
	p := state.Personality
	hasPers := state.HasPersonality

	var cands []candidate

	// ── 求職：安定需求 + 生存壓力共同驅動；秩序感高者更傾向找正職 ──
	if len(ctx.VenueRoomIDsInRange) > 0 {
		score := (stability + survival*0.5)
		if hasPers {
			score *= 0.5 + p.Orderliness*0.5
		}
		if score > 0.05 {
			cands = append(cands, candidate{
				intent: Intent{Type: IntentSeekJob, TargetRoomID: pickRandom(ctx.VenueRoomIDsInRange)},
				weight: score,
			})
		}
	}

	// ── 乞討：生存壓力驅動；膽量低者更願意乞討，敏感度高者更能放下面子 ──
	if hasTag(ctx.RoomTags, "gate", "social") || len(ctx.SocialOrGateInRange) > 0 {
		score := survival
		if hasPers {
			score *= (1.0 - p.Boldness*0.5) * (0.5 + p.Sensitivity*0.5)
		}
		if score > 0.05 {
			target := ""
			if !hasTag(ctx.RoomTags, "gate", "social") {
				target = pickRandom(ctx.SocialOrGateInRange)
			}
			cands = append(cands, candidate{
				intent: Intent{Type: IntentBeg, TargetRoomID: target},
				weight: score,
			})
		}
	}

	// ── 採集：生存壓力驅動；膽量高者更願意去荒野，秩序感低者更自由 ──
	if len(ctx.WildernessOutdoorInRange) > 0 {
		score := survival
		if hasPers {
			score *= (0.5 + p.Boldness*0.5) * (1.0 - p.Orderliness*0.5)
		}
		if score > 0.05 {
			cands = append(cands, candidate{
				intent: Intent{Type: IntentGather, TargetRoomID: pickRandom(ctx.WildernessOutdoorInRange)},
				weight: score,
			})
		}
	}

	// ── 社交：長時未交談 + 心境正面時驅動；敏感度高者更渴望社交 ──
	if len(ctx.TavernInRange) > 0 && social > 0 {
		score := social
		if hasPers {
			score *= 0.5 + p.Sensitivity*0.5
		}
		if score > 0.05 {
			target := ""
			if !hasTag(ctx.RoomTags, "tavern", "inn", "social") {
				target = pickRandom(ctx.TavernInRange)
			}
			cands = append(cands, candidate{
				intent: Intent{Type: IntentSocialize, TargetRoomID: target},
				weight: score,
			})
		}
	}

	// ── 自我實現：有工作且鎂足夠，依個性選額外活動 ──
	if state.HasAssignment && survival == 0 {
		if hasPers && p.Boldness > 0.6 && len(ctx.WildernessOutdoorInRange) > 0 {
			cands = append(cands, candidate{
				intent: Intent{Type: IntentGather, TargetRoomID: pickRandom(ctx.WildernessOutdoorInRange)},
				weight: 0.3 * p.Boldness,
			})
		}
		if hasPers && p.Orderliness > 0.6 && len(ctx.VenueRoomIDsInRange) > 0 {
			cands = append(cands, candidate{
				intent: Intent{Type: IntentWork, TargetRoomID: ctx.VenueRoomIDsInRange[0]},
				weight: 0.3 * p.Orderliness,
			})
		}
	}

	// ── 閒逛：底噪，始終存在；秩序感低者更常閒逛 ──
	wanderScore := 0.4
	if hasPers {
		wanderScore *= 1.5 - p.Orderliness
	}
	cands = append(cands, candidate{intent: Intent{Type: IntentWander}, weight: wanderScore})

	return weightedRandomSelect(cands)
}

// urgencySurvival 生存緊迫度：鎂越低越緊迫，>= 閾值為 0。
func urgencySurvival(mg int) float64 {
	if mg >= SurvivalLineMg {
		return 0
	}
	v := 1.0 - (float64(mg) / float64(SurvivalLineMg))
	if v < 0 {
		return 0
	}
	if v > 1 {
		return 1
	}
	return v
}

// urgencyStability 安定緊迫度：無指派者有高度不安感（0.8）；有指派則為 0。
func urgencyStability(hasAssignment bool) float64 {
	if hasAssignment {
		return 0
	}
	return 0.8
}

// urgencySocial 社交緊迫度：長時未交談漸升至 1.0；心境越低越不想出門。
// disposition [-100,+100] → 心境因子 [0, 1]；心境=0 → 0.5，心境=+100 → 1.0，心境=-100 → 0.0。
func urgencySocial(lastTalkAt int64, disposition int) float64 {
	if lastTalkAt <= 0 {
		return 0 // 從未交談：未知孤獨，不驅動社交
	}
	elapsed := time.Now().Unix() - lastTalkAt
	if elapsed < 0 {
		elapsed = 0
	}
	u := float64(elapsed) / socialDecaySec
	if u > 1 {
		u = 1
	}
	// disposition factor: [-100,+100] → [0.0, 1.0]
	dispFactor := (float64(disposition) + 100.0) / 200.0
	return u * dispFactor
}

// weightedRandomSelect 從有分數的候選清單中加權隨機選一意圖。
func weightedRandomSelect(cands []candidate) Intent {
	if len(cands) == 0 {
		return Intent{Type: IntentWander}
	}
	total := 0.0
	for _, c := range cands {
		total += c.weight
	}
	if total <= 0 {
		return cands[0].intent
	}
	r := rand.Float64() * total
	for _, c := range cands {
		r -= c.weight
		if r <= 0 {
			return c.intent
		}
	}
	return cands[len(cands)-1].intent
}

// ResolveBrainPath 依意圖解析出尋路目標房間（供 MoveBrain 使用）。若意圖已帶 TargetRoomID 則用該房；否則依類型用 ctx 或圖找目標。
func ResolveBrainPath(g *RoomGraph, intent Intent, currentRoomID string, ctx *DecisionContext, maxDist int) []string {
	if intent.TargetRoomID != "" && intent.TargetRoomID != currentRoomID {
		path := g.FindPath(currentRoomID, intent.TargetRoomID)
		if len(path) > 0 {
			return path
		}
	}
	if maxDist <= 0 {
		maxDist = 25
	}
	switch intent.Type {
	case IntentSeekJob:
		if ctx != nil && len(ctx.VenueRoomIDsInRange) > 0 {
			target := pickRandom(ctx.VenueRoomIDsInRange)
			if target != currentRoomID {
				if path := g.FindPath(currentRoomID, target); len(path) > 0 {
					return path
				}
			}
		}
		room, _ := g.FindNearestByTag(currentRoomID, "social", maxDist)
		if room != "" && room != currentRoomID {
			return g.FindPath(currentRoomID, room)
		}
	case IntentBeg:
		room, _ := g.FindNearestByTag(currentRoomID, "gate", maxDist)
		if room == "" {
			room, _ = g.FindNearestByTag(currentRoomID, "social", maxDist)
		}
		if room != "" && room != currentRoomID {
			return g.FindPath(currentRoomID, room)
		}
	case IntentGather:
		room, _ := g.FindNearestByTag(currentRoomID, "wilderness", maxDist)
		if room == "" {
			room, _ = g.FindNearestByTag(currentRoomID, "outdoor", maxDist)
		}
		if room != "" && room != currentRoomID {
			return g.FindPath(currentRoomID, room)
		}
	case IntentWork:
		if ctx != nil && len(ctx.VenueRoomIDsInRange) > 0 {
			target := ctx.VenueRoomIDsInRange[0]
			if target != currentRoomID {
				if path := g.FindPath(currentRoomID, target); len(path) > 0 {
					return path
				}
			}
		}
	case IntentSocialize:
		if ctx != nil && len(ctx.TavernInRange) > 0 {
			target := pickRandom(ctx.TavernInRange)
			if target != currentRoomID {
				if path := g.FindPath(currentRoomID, target); len(path) > 0 {
					return path
				}
			}
		}
		for _, tag := range []string{"tavern", "inn", "social"} {
			room, _ := g.FindNearestByTag(currentRoomID, tag, maxDist)
			if room != "" && room != currentRoomID {
				return g.FindPath(currentRoomID, room)
			}
		}
	case IntentWander:
		candidates := g.FindRoomsWithinDist(currentRoomID, []string{"social", "gate", "outdoor"}, maxDist)
		var filtered []string
		for _, c := range candidates {
			if c != currentRoomID {
				filtered = append(filtered, c)
			}
		}
		if len(filtered) > 0 {
			target := filtered[rand.Intn(len(filtered))]
			return g.FindPath(currentRoomID, target)
		}
	}
	return nil
}

File: src/npc/decision.rs
//! NPC 決策引擎（對齊 Go `npc/decision.go`）：`Decide` 產生意圖，`ResolveBrainPath` 轉成尋路序列。
// 結構刻意鏡像 Go 之多層判斷，合併 if 反而難與原版對照。
#![allow(clippy::collapsible_if)]

use std::time::{SystemTime, UNIX_EPOCH};

use rand::Rng;

use crate::db::{
    expand_soul_seed_to_personality, get_all_venue_room_ids, get_assignments_for_entity, get_entity,
    get_recent_events, inventory_total_qty, Personality, RoomGraph, EVT_TALK,
};

/// 生存層鎂閾值（對齊 Go `SurvivalLineMg`）。
pub const SURVIVAL_LINE_MG: i32 = 50;
/// 社交衰減週期秒（對齊 Go `socialDecaySec`）。
const SOCIAL_DECAY_SEC: i64 = 7200;

/// 意圖類型（對齊 Go `IntentType`）。
#[derive(Debug, Clone, Copy, PartialEq, Eq, Default)]
pub enum IntentType {
    SeekJob,
    Beg,
    Trade,
    Gather,
    #[default]
    Wander,
    Work,
    Idle,
    Socialize,
}

impl IntentType {
    /// 對應 `gametext.json` brain_arrival 鍵與事件語意。
    #[must_use]
    pub fn as_key(&self) -> &'static str {
        match self {
            Self::SeekJob => "seek_job",
            Self::Beg => "beg",
            Self::Trade => "trade",
            Self::Gather => "gather",
            Self::Wander => "wander",
            Self::Work => "work",
            Self::Idle => "idle",
            Self::Socialize => "socialize",
        }
    }
}

/// 決策輸出：單一意圖與可選目標房（對齊 Go `Intent`）。
#[derive(Debug, Clone, Default)]
pub struct Intent {
    pub ty: IntentType,
    pub target_room_id: String,
}

/// 決策輸入：NPC 自身狀態（對齊 Go `DecisionState`）。
#[derive(Debug, Clone)]
pub struct DecisionState {
    pub magnesium: i32,
    pub has_assignment: bool,
    pub personality: Personality,
    pub has_personality: bool,
    pub last_talk_at: i64,
    pub disposition: i32,
    pub inventory_total_qty: i32,
}

/// 世界情境（對齊 Go `DecisionContext`）。
#[derive(Debug, Clone, Default)]
pub struct DecisionContext {
    pub room_tags: Vec<String>,
    pub venue_room_ids_in_range: Vec<String>,
    pub social_or_gate_in_range: Vec<String>,
    pub wilderness_outdoor_in_range: Vec<String>,
    pub tavern_in_range: Vec<String>,
}

struct Candidate {
    intent: Intent,
    weight: f64,
}

/// 由 store 組裝決策狀態（對齊 Go `BuildDecisionState`）。
pub fn build_decision_state(entity_id: &str) -> anyhow::Result<DecisionState> {
    let Some(ch) = get_entity(entity_id)? else {
        anyhow::bail!("entity not found");
    };
    let assignments = get_assignments_for_entity(entity_id)?;
    let has_assignment = !assignments.is_empty();
    let (personality, has_personality) = if let Some(seed) = ch.soul_seed {
        (expand_soul_seed_to_personality(seed), true)
    } else {
        (Personality {
            boldness: 0.5,
            sensitivity: 0.5,
            orderliness: 0.5,
        }, false)
    };
    let inv_qty = inventory_total_qty(&ch.inventory);
    let mut last_talk_at: i64 = 0;
    for e in get_recent_events(entity_id, 10) {
        if e.event_type == EVT_TALK {
            last_talk_at = e.at;
            break;
        }
    }
    Ok(DecisionState {
        magnesium: ch.magnesium,
        has_assignment,
        personality,
        has_personality,
        last_talk_at,
        disposition: ch.disposition,
        inventory_total_qty: inv_qty,
    })
}

/// 組裝周邊情境（對齊 Go `BuildDecisionContext`）。
pub fn build_decision_context(g: &RoomGraph, current_room_id: &str, max_dist: i32) -> DecisionContext {
    let mut md = max_dist;
    if md <= 0 {
        md = 20;
    }
    let venue_rooms = get_all_venue_room_ids().unwrap_or_default();
    DecisionContext {
        room_tags: g.room_tags(current_room_id),
        venue_room_ids_in_range: filter_rooms_within_dist(g, current_room_id, &venue_rooms, md),
        social_or_gate_in_range: g.find_rooms_within_dist(current_room_id, &["social", "gate"], md),
        wilderness_outdoor_in_range: g.find_rooms_within_dist(
            current_room_id,
            &["wilderness", "outdoor"],
            md,
        ),
        tavern_in_range: g.find_rooms_within_dist(current_room_id, &["tavern", "inn"], md),
    }
}

fn filter_rooms_within_dist(g: &RoomGraph, origin: &str, candidates: &[String], max_dist: i32) -> Vec<String> {
    if candidates.is_empty() || max_dist <= 0 {
        return Vec::new();
    }
    let mut out: Vec<String> = Vec::new();
    for c in candidates {
        if c == origin {
            out.push(c.clone());
            continue;
        }
        if let Some(p) = g.find_path(origin, c) {
            if !p.is_empty() && p.len() <= max_dist as usize {
                out.push(c.clone());
            }
        }
    }
    out
}

fn has_tag(tags: &[String], want: &[&str]) -> bool {
    for t in tags {
        for w in want {
            if t == w {
                return true;
            }
        }
    }
    false
}

fn pick_random<T: Clone>(xs: &[T]) -> T {
    let mut rng = rand::rng();
    xs[rng.random_range(0..xs.len())].clone()
}

fn urgency_survival(mg: i32) -> f64 {
    if mg >= SURVIVAL_LINE_MG {
        return 0.0;
    }
    let v = 1.0 - (f64::from(mg) / f64::from(SURVIVAL_LINE_MG));
    v.clamp(0.0, 1.0)
}

fn urgency_stability(has_assignment: bool) -> f64 {
    if has_assignment {
        0.0
    } else {
        0.8
    }
}

fn urgency_social(last_talk_at: i64, disposition: i32) -> f64 {
    if last_talk_at <= 0 {
        return 0.0;
    }
    let now = SystemTime::now()
        .duration_since(UNIX_EPOCH)
        .map(|d| d.as_secs() as i64)
        .unwrap_or(0);
    let elapsed = now.saturating_sub(last_talk_at).max(0);
    let u = (elapsed as f64 / SOCIAL_DECAY_SEC as f64).min(1.0);
    let disp_factor = (f64::from(disposition) + 100.0) / 200.0;
    u * disp_factor
}

fn weighted_random_select(cands: &[Candidate]) -> Intent {
    if cands.is_empty() {
        return Intent {
            ty: IntentType::Wander,
            target_room_id: String::new(),
        };
    }
    let total: f64 = cands.iter().map(|c| c.weight).sum();
    if total <= 0.0 {
        return cands[0].intent.clone();
    }
    let mut rng = rand::rng();
    let mut r = rng.random_range(0.0..total);
    for c in cands {
        r -= c.weight;
        if r <= 0.0 {
            return c.intent.clone();
        }
    }
    cands.last().expect("cands non-empty").intent.clone()
}

/// 評分引擎：加權隨機選意圖（對齊 Go `Decide`）。
pub fn decide(state: DecisionState, ctx: &DecisionContext) -> Intent {
    let survival = urgency_survival(state.magnesium);
    let stability = urgency_stability(state.has_assignment);
    let social = urgency_social(state.last_talk_at, state.disposition);
    let p = state.personality;
    let has_pers = state.has_personality;
    let mut cands: Vec<Candidate> = Vec::new();

    if !ctx.venue_room_ids_in_range.is_empty() {
        let mut score = stability + survival * 0.5;
        if has_pers {
            score *= 0.5 + p.orderliness * 0.5;
        }
        if score > 0.05 {
            cands.push(Candidate {
                intent: Intent {
                    ty: IntentType::SeekJob,
                    target_room_id: pick_random(&ctx.venue_room_ids_in_range),
                },
                weight: score,
            });
        }
    }

    if has_tag(&ctx.room_tags, &["gate", "social"]) || !ctx.social_or_gate_in_range.is_empty() {
        let mut score = survival;
        if has_pers {
            score *= (1.0 - p.boldness * 0.5) * (0.5 + p.sensitivity * 0.5);
        }
        if score > 0.05 {
            let target = if !has_tag(&ctx.room_tags, &["gate", "social"]) {
                pick_random(&ctx.social_or_gate_in_range)
            } else {
                String::new()
            };
            cands.push(Candidate {
                intent: Intent {
                    ty: IntentType::Beg,
                    target_room_id: target,
                },
                weight: score,
            });
        }
    }

    if !ctx.wilderness_outdoor_in_range.is_empty() {
        let mut score = survival;
        if has_pers {
            score *= (0.5 + p.boldness * 0.5) * (1.0 - p.orderliness * 0.5);
        }
        if score > 0.05 {
            cands.push(Candidate {
                intent: Intent {
                    ty: IntentType::Gather,
                    target_room_id: pick_random(&ctx.wilderness_outdoor_in_range),
                },
                weight: score,
            });
        }
    }

    if state.inventory_total_qty > 0 && survival > 0.05
        && (!ctx.social_or_gate_in_range.is_empty() || has_tag(&ctx.room_tags, &["gate", "social"]))
    {
        let mut score = survival * 0.45;
        if has_pers {
            score *= 0.5 + p.boldness * 0.5;
        }
        if score > 0.05 {
            let target = if !has_tag(&ctx.room_tags, &["gate", "social"]) {
                pick_random(&ctx.social_or_gate_in_range)
            } else {
                String::new()
            };
            cands.push(Candidate {
                intent: Intent {
                    ty: IntentType::Trade,
                    target_room_id: target,
                },
                weight: score,
            });
        }
    }

    if !ctx.tavern_in_range.is_empty() && social > 0.0 {
        let mut score = social;
        if has_pers {
            score *= 0.5 + p.sensitivity * 0.5;
        }
        if score > 0.05 {
            let target = if !has_tag(&ctx.room_tags, &["tavern", "inn", "social"]) {
                pick_random(&ctx.tavern_in_range)
            } else {
                String::new()
            };
            cands.push(Candidate {
                intent: Intent {
                    ty: IntentType::Socialize,
                    target_room_id: target,
                },
                weight: score,
            });
        }
    }

    if state.has_assignment && survival == 0.0 {
        if has_pers && p.boldness > 0.6 && !ctx.wilderness_outdoor_in_range.is_empty() {
            cands.push(Candidate {
                intent: Intent {
                    ty: IntentType::Gather,
                    target_room_id: pick_random(&ctx.wilderness_outdoor_in_range),
                },
                weight: 0.3 * p.boldness,
            });
        }
        if has_pers && p.orderliness > 0.6 && !ctx.venue_room_ids_in_range.is_empty() {
            cands.push(Candidate {
                intent: Intent {
                    ty: IntentType::Work,
                    target_room_id: ctx.venue_room_ids_in_range[0].clone(),
                },
                weight: 0.3 * p.orderliness,
            });
        }
    }

    let mut wander_score = 0.4;
    if has_pers {
        wander_score *= 1.5 - p.orderliness;
    }
    cands.push(Candidate {
        intent: Intent {
            ty: IntentType::Wander,
            target_room_id: String::new(),
        },
        weight: wander_score,
    });

    weighted_random_select(&cands)
}

/// 依意圖解析尋路目標序列（對齊 Go `ResolveBrainPath`）。
pub fn resolve_brain_path(
    g: &RoomGraph,
    intent: &Intent,
    current_room_id: &str,
    ctx: &DecisionContext,
    max_dist: i32,
) -> Vec<String> {
    let mut md = max_dist;
    if md <= 0 {
        md = 25;
    }
    if !intent.target_room_id.is_empty() && intent.target_room_id != current_room_id {
        if let Some(path) = g.find_path(current_room_id, &intent.target_room_id) {
            if !path.is_empty() {
                return path;
            }
        }
    }
    match intent.ty {
        IntentType::SeekJob => {
            if !ctx.venue_room_ids_in_range.is_empty() {
                let target = pick_random(&ctx.venue_room_ids_in_range);
                if target != current_room_id {
                    if let Some(p) = g.find_path(current_room_id, &target) {
                        if !p.is_empty() {
                            return p;
                        }
                    }
                }
            }
            if let Some((room, _)) = g.find_nearest_by_tag(current_room_id, "social", md) {
                if room != current_room_id {
                    if let Some(p) = g.find_path(current_room_id, &room) {
                        return p;
                    }
                }
            }
        }
        IntentType::Beg => {
            let room = g
                .find_nearest_by_tag(current_room_id, "gate", md)
                .map(|(r, _)| r)
                .filter(|r| r != current_room_id)
                .or_else(|| {
                    g.find_nearest_by_tag(current_room_id, "social", md)
                        .map(|(r, _)| r)
                        .filter(|r| r != current_room_id)
                });
            if let Some(r) = room {
                if let Some(p) = g.find_path(current_room_id, &r) {
                    return p;
                }
            }
        }
        IntentType::Trade => {
            let room = g
                .find_nearest_by_tag(current_room_id, "social", md)
                .map(|(r, _)| r)
                .filter(|r| r != current_room_id)
                .or_else(|| {
                    g.find_nearest_by_tag(current_room_id, "gate", md)
                        .map(|(r, _)| r)
                        .filter(|r| r != current_room_id)
                });
            if let Some(r) = room {
                if let Some(p) = g.find_path(current_room_id, &r) {
                    return p;
                }
            }
        }
        IntentType::Gather => {
            let room = g
                .find_nearest_by_tag(current_room_id, "wilderness", md)
                .map(|(r, _)| r)
                .filter(|r| r != current_room_id)
                .or_else(|| {
                    g.find_nearest_by_tag(current_room_id, "outdoor", md)
                        .map(|(r, _)| r)
                        .filter(|r| r != current_room_id)
                });
            if let Some(r) = room {
                if let Some(p) = g.find_path(current_room_id, &r) {
                    return p;
                }
            }
        }
        IntentType::Work => {
            if !ctx.venue_room_ids_in_range.is_empty() {
                let target = ctx.venue_room_ids_in_range[0].clone();
                if target != current_room_id {
                    if let Some(p) = g.find_path(current_room_id, &target) {
                        if !p.is_empty() {
                            return p;
                        }
                    }
                }
            }
        }
        IntentType::Socialize => {
            if !ctx.tavern_in_range.is_empty() {
                let target = pick_random(&ctx.tavern_in_range);
                if target != current_room_id {
                    if let Some(p) = g.find_path(current_room_id, &target) {
                        if !p.is_empty() {
                            return p;
                        }
                    }
                }
            }
            for tag in ["tavern", "inn", "social"] {
                if let Some((room, _)) = g.find_nearest_by_tag(current_room_id, tag, md) {
                    if room != current_room_id {
                        if let Some(p) = g.find_path(current_room_id, &room) {
                            return p;
                        }
                    }
                }
            }
        }
        IntentType::Wander => {
            let candidates = g.find_rooms_within_dist(current_room_id, &["social", "gate", "outdoor"], md);
            let filtered: Vec<String> = candidates.into_iter().filter(|c| c != current_room_id).collect();
            if !filtered.is_empty() {
                let target = pick_random(&filtered);
                if let Some(p) = g.find_path(current_room_id, &target) {
                    return p;
                }
            }
        }
        IntentType::Idle => {}
    }
    Vec::new()
}
